Feature expose provider logs

LucasRoesler opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This creates a new logs endpoint that proxies the http request to the log provider as described in openfaas/faas#751


## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

See openfaas/faas#751

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

This has been manually tested as well as some initial unit tests have been added to verify that it does not leak any goroutines 

This can be tested using these instructions 
- https://gist.github.com/LucasRoesler/a6d0c5e2e669825ec4ab613e3487f1e8 (installs with the updated k8s provider)
 - https://gist.github.com/LucasRoesler/ec0c365d02296c29e66be683247be8aa

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

alexellis commented on gateway/handlers/logs.go:
```diff
@@ -0,0 +1,133 @@
+		}
+
+		defer clientConn.Close()
+		clientConn.SetWriteDeadline(time.Time{}) // allow arbitrary time between log writes
```
Self document that `time.Time{}` (empty time value) means there is no `WriteDeadline`.

LucasRoesler replied:
Doc added

alexellis commented on gateway/handlers/logs.go:
```diff
@@ -0,0 +1,133 @@
+
+		// write response headers directly to the buffer per RFC 2616
+		// https://www.w3.org/Protocols/rfc2616/rfc2616-sec6.html
+		buf.WriteString("HTTP/1.1 " + logResp.Status + crlf)
```
I think there's a const for that in the `http` package, not sure if it's public, worth checking.

LucasRoesler replied:
There is no public constant documented in the http package

alexellis commented on gateway/handlers/logs.go:
```diff
@@ -0,0 +1,133 @@
+			buf.Flush()
+			return
+		}
+		defer logResp.Body.Close()
```
?
```
if logResp.Body !=nil {

}
``

LucasRoesler replied:
I have added a comment that the nil check is not needed

alexellis commented on gateway/handlers/forwarding_proxy.go:
```diff
@@ -139,6 +139,10 @@ func copyHeaders(destination http.Header, source *http.Header) {
 		(destination)[k] = vClone
 	}
+
+	for _, h := range hopHeaders {
```
@stefanprodan PTAL

alexellis replied:
Removed here -> https://github.com/openfaas/faas/pull/1146

alexellis replied:
@LucasRoesler we added this code elsewhere please can you remove this code if duplicated?

alexellis commented on gateway/handlers/logs.go:
```diff
@@ -0,0 +1,133 @@
+const upstreamLogsEndpoint = "/system/logs"
+
+// NewLogHandlerFunc creates and http HandlerFunc from the supplied log Requestor.
+func NewLogHandlerFunc(logProvider url.URL) http.HandlerFunc {
```
@stefanprodan This is what code looks like when WriteTimeout is enabled on the http.Server.

LucasRoesler commented:
@alexellis i have updated and validated the proxy implementation using only CloseNotifier and Flusher, I also rebased it on master to handle the merge conflict. I have not removed the write timeout yet, but I have validated that it works 100% as expected when you do remove the write timeout.

The license check is failing on the test utility I vendored to ensure the proxy endpoint does not cause any goroutine leaks.

alexellis commented on gateway/handlers/logs.go:
```diff
@@ -0,0 +1,225 @@
+const upstreamLogsEndpoint = "/system/logs"
+
+// NewLogHandlerFunc creates and http HandlerFunc from the supplied log Requestor.
+func NewLogHandlerFunc(logProvider url.URL) http.HandlerFunc {
```
This is where we are using stdlib

alexellis commented on gateway/handlers/logs.go:
```diff
@@ -0,0 +1,225 @@
+
+// NewHijackLogHandlerFunc creates and http HandlerFunc from the supplied log Requestor.
+// This remains in the package for reference purposes, providers should instead use NewLogHandlerFunc
+func NewHijackLogHandlerFunc(logProvider url.URL) http.HandlerFunc {
```
Taking over the connection.

alexellis replied:
Can you put this into a gist or file somewhere outside the PR?

alexellis commented on gateway/handlers/logs.go:
```diff
@@ -0,0 +1,225 @@
+
+// NewLogHandlerFunc creates and http HandlerFunc from the supplied log Requestor.
+func NewLogHandlerFunc(logProvider url.URL) http.HandlerFunc {
+	writeRequestURI := false
```
Initial version - bring in a time.Duration and kill the connection using `time.AfterFunc()`

LucasRoesler replied:
This was implemented with timeout, as in the the next comment

alexellis commented on gateway/handlers/logs.go:
```diff
@@ -0,0 +1,225 @@
+	upstreamLogProviderBase := strings.TrimSuffix(logProvider.String(), "/")
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
```
or `WithTimeout()`

alexellis commented:
If the provider gives 404, we should return Not implemented to the CLI and handle that. i.e. the provider may not support logs message.

LucasRoesler commented:
@alexellis i have updated this but haven't had a chance to fully test it because license check needs a new release

LucasRoesler commented:
With my most recent commit, i have pushed a public tag of this  gateway to `theaxer/gateway:loggerhandler` and if you use the the updates on the CLI PR, you can now use these instructions to test the logs 

```
### You need 3 terminals

## Terminal 1

### You need the logs branch of the faas-cli
cd $PATH_TO_FAAS_CLI_REPO
hub pr checkout 613
make local-install

### Now go to your faas-netes repo
cd $PATH_TO_FAAS_NETES_REPO

make start-kind

cat password.txt | faas-cli login --gateway=http://localhost:31112 -u admin -s

export KUBECONFIG=$(kind get kubeconfig-path --name="${OF_DEV_ENV:-kind}")

helm upgrade openfaas --install ./chart/openfaas \
    --namespace openfaas  \
    --set basic_auth=true \
    --set openfaasImagePullPolicy=IfNotPresent \
    --set faasnetes.imagePullPolicy=IfNotPresent \
    --set gateway.image=theaxer/gateway:loggerhandler \
    --set faasnetes.image=openfaas/faas-netes:0.8.0 \
    --set functionNamespace=openfaas-fn

bash ./contrib/restart_port_forward.sh

## Terminal 2
faas-cli logs echo --gateway=http://localhost:31112

## Terminal 3
echo "`date`"  | faas-cli invoke echo --gateway=http://localhost:31112
```

LucasRoesler commented:
@alexellis all of the comments should be addressed nwo

alexellis commented:
Please can you run us through the decision to add `goleak` from Uber?

I.e.

* what it does?
* why we need it?
* if we need it here, where else do we need it?

alexellis commented:
Please can you document `logs_provider_url` and any other configuration items in the gateway/README.md and in any other places you feel appropriate?

alexellis commented:
Please can you add a unit test for readconfig for the new `LogsProviderURL` value? Use the existing tests as an example, generally it's something like: default (not set) / override and n/a.

alexellis commented on gateway/Gopkg.toml:
```diff
@@ -27,3 +27,7 @@ ignored = ["github.com/openfaas/faas/gateway/queue"]
 [[constraint]]
   name = "github.com/openfaas/faas-provider"
   version = "0.9.1"
```
Would you be able to vendor 0.9.2 - the latest version? https://github.com/openfaas/faas-provider/releases/tag/0.9.2

LucasRoesler commented:
@alexellis, i have updated the PR to address your comments. 

with respect to the questions about `goleak`, `goleak.VerifyNoLeaks` will count the goroutines before and after the test is run to ensure that there are no stray goroutines started by the code.  This is important for the log handler because the code starts at least one goroutine to handle the streaming the logs in the background, this is one way to help ensure that the logs handler is not leaking memory. Usage is strictly adding defer `goleak.VerifyNoLeaks(t)` to the start of the test.

alexellis commented:
Given that we've adopted `goleak`, please can you document its usage in the CONTRIBUTING file? If you don't want this to be a mega-PR then raise an issue to track it now and it can be done later

alexellis commented:
As per https://github.com/openfaas/faas/pull/1130#issuecomment-506805963 - let's document the choices we are making

alexellis commented on gateway/README.md:
```diff
@@ -45,7 +45,9 @@ See the [openfaas/store](https://github.com/openfaas/store) repo for more.
 
 ## Logs
 
-Logs are available at the function level and can be accessed through Swarm or Kubernetes using native tooling. You can also install a Docker logging driver to aggregate your logs. By default functions will not write the request and response bodies to stdout. You can toggle this behaviour by setting `read_debug` for the request and `write_debug` for the response.
```
The note on a logging aggregation tool is still valid, I'm not sure that should be deleted.
